consensus/src/consensusdb/mod.rs

```diff
@@ -55,6 +55,9 @@ impl ConsensusDB {
     )> {
         let last_vote = self.get_last_vote()?;
         let highest_timeout_certificate = self.get_highest_timeout_certificate()?;
+        // @REVIEW: seems that we don't record the executed state ID of the block in storage. So
+        // in theory, a block without a QC could be executed with state hash X, then after restart
+        // (due to an error) we recompute it with state Y.
```

Comment:
it's intentional, if it has a QC, recompute would ensure it's the same otherwise panic, if it doesn't have QC, we'll never equivocate because of TCB.
I was hoping this helps in case we see a QC which doesn't match local compute (due to an error) and we panic, after restart recompute can fix the error and recover.

Reply:
Let's say you get a block B1, you execute it and it has state S1. You vote on that state. Then you crash before a QC is formed on B1.

Later, you get a block B2 that build on B1. You execute B as part of computing B' and get a state S1'. When you vote on B2, you'll implicitly be endorsing S1' even though you voted on S1. Strictly speaking, not an equivocation / you aren't doing multiple votes per round... but not sure if it's the right behavior.

I realize this is a corner case that only happens with non-deterministic execution

Reply:
it's true in the sense of transitive result, but not sure if it violates safety since the node still only votes once per round.